src/rust/cryptography-x509/src/extensions.rs

```diff
@@ -2,16 +2,58 @@
 // 2.0, and the BSD License. See the LICENSE file in the root of this repository
 // for complete details.
 
+use std::collections::HashSet;
+
 use crate::common;
 use crate::crl;
 use crate::name;
 
-pub type Extensions<'a> = common::Asn1ReadableOrWritable<
+pub type RawExtensions<'a> = common::Asn1ReadableOrWritable<
     'a,
     asn1::SequenceOf<'a, Extension<'a>>,
     asn1::SequenceOfWriter<'a, Extension<'a>, Vec<Extension<'a>>>,
 >;
 
+pub struct Extensions<'a> {
+    inner: RawExtensions<'a>,
+}
+
+impl<'a> Extensions<'a> {
+    pub fn from_raw_extensions(
+        raw: Option<&RawExtensions<'a>>,
+    ) -> Result<Option<Self>, asn1::ObjectIdentifier> {
```

Comment:
I'm not super happy with this signature -- we end up pushing an `Option` through as a no-op to avoid having to destructure the optional extensions in multiple other places, but that results in more invalid states than are strictly needed here (and the smelly `Result<Option<...>>` type).

Reply:
I think a reasonable follow up is, instead of `as_raw()`, we make `Extensions` into an iterator, and just handle the `Option<>` internally. But let's do that in a seperate PR, this one is already complicated enough.